arch/risc-v/src/mpfs/mpfs_fpga_canfd.c

```diff
@@ -0,0 +1,2917 @@
+
+          switch (txb_status)
+            {
+            case TXT_TOK:
+              break;
+            case TXT_ERR:
+              canwarn("TXB in Error state\n");
+              break;
+            case TXT_ABT:
+              canwarn("TXB in Aborted state\n");
+              break;
+            default:
+              other_status = true;
+
+              if (first_buffer)
+                {
+                  canerr("TXB#%u not in a finished state (0x%x)!\n",
+                         txb_id, txb_status);
+
+                  priv->txb_processed++;
+
+                  /* Rotate TXT buffer priority */
+
+                  mpfs_can_update_txb_prio(priv);
+
+                  /* Mark current unfinished state TXT buffer as empty */
+
+                  mpfs_can_send_txb_cmd(priv, TXT_CMD_SET_EMPTY, txb_id);
+
+                  /* Something is fishy. CLear txb status change interrupt */
+
+                  putreg32(MPFS_CANFD_INT_STAT_TXBHCI,
+                           priv->base + MPFS_CANFD_INT_STAT_OFFSET);
+
+                  return;
+                }
+              break;
+            }
+
+          if (other_status)
+            {
+              break;
+            }
+          else
+            {
+              priv->txb_processed++;
+              first_buffer = false;
+              buffer_processed = true;
+
+              /* Rotate TXT buffer priority */
+
+              mpfs_can_update_txb_prio(priv);
+
+              /* Mark current finished state TXT buffer as empty */
+
+              mpfs_can_send_txb_cmd(priv, TXT_CMD_SET_EMPTY, txb_id);
+            }
+        }
+
+      if (buffer_processed)
+        {
+          /* Since there are some buffers processed and the number of TXT
+           * used now matched the number of processed buffer, we can clear
+           * the interrupt in order to avoid any erroneous interrupt after
+           * the last true TXT buffer interrupt is processed.
+           */
+
+          putreg32(MPFS_CANFD_INT_STAT_TXBHCI,
+                   priv->base + MPFS_CANFD_INT_STAT_OFFSET);
+        }
+    }
+  while (buffer_processed);
+}
+
+/****************************************************************************
+ * Name: mpfs_txdone_work
+ *
+ * Description:
+ *  An interrupt was received indicating that the last TX packet(s) is done
+ *
+ * Input Parameters:
+ *  priv  - Pointer to the private CAN driver state structure
+ *
+ * Returned Value:
+ *  None
+ *
+ * Assumptions:
+ *  Global interrupts are disabled by the watchdog logic.
+ *  We are not in an interrupt context so that we can lock the network.
+ *
+ ****************************************************************************/
+
+static void mpfs_txdone_work(void *arg)
+{
+  struct mpfs_driver_s *priv = (struct mpfs_driver_s *)arg;
+
+  /* There should be space for a new TX in any event.  Poll the network for
+   * new XMIT data
+   */
+
+  net_lock();
+  devif_poll(&priv->dev, mpfs_txpoll);
```

Comment:
> Is this really how the net <-> driver communication is supposed to work? The driver polls the interface if there is more work to do?

-> yes. Once an txt buffer is processed (packet sent), driver can poll the network for new TX if there is any. This is one of two ways new TX can be initiated. The other way is via direct callback via `priv->dev.d_txavail = mpfs_txavail`